src/sampler.jl

```diff
@@ -0,0 +1,60 @@
+#=
+    `MVNSampler` is used to draw from multivariate normal distribution
+=#
+
+import Base: ==
+
+immutable MVNSampler{TM<:Real,TS<:Real,TQ<:LinAlg.BlasReal}
+    mu::Vector{TM}
+    Sigma::Matrix{TS}
+    Q::Matrix{TQ}
+end
+
+function MVNSampler{TM<:Real,TS<:Real}(mu::Vector{TM}, Sigma::Matrix{TS})
+    ATOL1, RTOL1 = 1e-8, 1e-8
+    ATOL2, RTOL2 = 1e-8, 1e-14
+
+    n = length(mu)
+
+    if size(Sigma) != (n,n) # Check Sigma is n x n
+        throw(ArgumentError("Sigma must be 2 dimensional and square matrix of same length to mu"))
+    end
+
+    issymmetric(Sigma) || throw(ArgumentError("Sigma must be symmetric"))
+
+    C = cholfact(Symmetric(Sigma, :L), Val{true})
+
+    if C.rank == n  # Positive definite
+        p = invperm(C.piv)
+        Q = tril(C.factors)[p,p]
+        return MVNSampler(mu, Sigma, Q)
+    end
+
+    non_PSD_msg = "Sigma must be positive semidefinite"
+
+    for i in C.rank+1:n
+        if C[:L][i, i] < -ATOL1 || C[:L][i, i]/norm(diag(C[:L])) < -RTOL1 # Not positive semidefinite

```

Comment:
Don't do division.

Try this:

```jl
        if C[:L][i, i] < -ATOL1 - RTOL1 * C[:L][1, 1]
```

(I have no logical justification, but propose to make it relative just to the largest diagonal element.)

You will save one line if you do this:

```jl
    for i in C.rank+1:n
        C[:L][i, i] >= -ATOL1 - RTOL1 * C[:L][1, 1] ||
            throw(ArgumentError(non_PSD_msg))
    end
```